### PR TITLE
SONARJNKNS-232 Adapt SonarQube configuration to server version

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/hudson/plugins/sonar/utils/SonarMaven.java [hunk 1]
```diff
@@ -112,7 +112,10 @@ protected void wrapUpArguments(ArgumentListBuilder args, String normalizedTarget
 
     argsBuilder.append("sonar.branch", publisher.getBranch());
 
-    if (StringUtils.isNotBlank(getInstallation().getSonarLogin())) {
+    if(StringUtils.isNotBlank(getInstallation().getServerAuthenticationToken())) {
+      argsBuilder.appendMasked("sonar.login", getInstallation().getServerAuthenticationToken());
+    }
+    else if (StringUtils.isNotBlank(getInstallation().getSonarLogin())) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/SonarMaven.java]: MINOR Move this "else" on the same line that the previous closing curly brace. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java [hunk 2]
```diff
@@ -52,7 +52,13 @@ public ProjectInformation resolve(String projectUrl, String ceTaskId, String ins
     try {
       String projectKey = extractProjectKey(projectUrl);
       String serverUrl = extractServerUrl(projectUrl);
-      WsClient wsClient = new WsClient(client, serverUrl, inst.getSonarLogin(), inst.getSonarPassword());
+      
+      WsClient wsClient;
+      if(StringUtils.isNotEmpty(inst.getServerAuthenticationToken())) {
+        wsClient = new WsClient(client, serverUrl, inst.getServerAuthenticationToken(), null);
+      } else {
+       wsClient = new WsClient(client, serverUrl, inst.getSonarLogin(), inst.getSonarPassword());
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: MAJOR Make this line start at column 9. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AIndentationCheck)

File: src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package hudson.plugins.sonar.utils;
+
+/**
+ * Uses plain Strings, serializing specific reference versions as values to make serialization and posterior modifications easy. 
+ */
+public class SQServerVersions {
+  public static String SQ_5_1_OR_LOWER = "5.1";
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java]: MAJOR Make this "public static SQ_5_1_OR_LOWER" field final [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1444)
MINOR Rename this field "SQ_5_1_OR_LOWER" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3008)
MAJOR Make SQ_5_1_OR_LOWER a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

File: src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package hudson.plugins.sonar.utils;
+
+/**
+ * Uses plain Strings, serializing specific reference versions as values to make serialization and posterior modifications easy. 
+ */
+public class SQServerVersions {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1118)

File: src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  public static String SQ_5_1_OR_LOWER = "5.1";
+  public static String SQ_5_2 = "5.2";
+  public static String SQ_5_3_OR_HIGHER = "5.3";
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java]: MAJOR Make this "public static SQ_5_3_OR_HIGHER" field final [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1444)
MINOR Rename this field "SQ_5_3_OR_HIGHER" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3008)
MAJOR Make SQ_5_3_OR_HIGHER a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

File: src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package hudson.plugins.sonar.utils;
+
+/**
+ * Uses plain Strings, serializing specific reference versions as values to make serialization and posterior modifications easy. 
+ */
+public class SQServerVersions {
+  public static String SQ_5_1_OR_LOWER = "5.1";
+  public static String SQ_5_2 = "5.2";
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/utils/SQServerVersions.java]: MAJOR Make this "public static SQ_5_2" field final [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1444)
MINOR Rename this field "SQ_5_2" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3008)
MAJOR Make SQ_5_2 a static final constant or non-public and provide accessors if needed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 13 issues:
* MAJOR 9 major
* MINOR 4 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/d8777b6ea21283cc6c454d161e3c2907d23de865/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by henryju** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/ce2c355406c5be37d4bf1566b0338c58b63f0800/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/139e45e1d7d87e06e94ace9056387121eca309a5/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/ae6f778fe1a9cf3b9b195d0fda8d30717adee972/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/ae6f778fe1a9cf3b9b195d0fda8d30717adee972/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/833c3c145970cde311672d9a9bad9d879c6197cd/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [The Cyclomatic Complexity of this method "resolve" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/833c3c145970cde311672d9a9bad9d879c6197cd/src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java#L45) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "populateConfiguration" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/833c3c145970cde311672d9a9bad9d879c6197cd/src/main/java/hudson/plugins/sonar/SonarRunnerBuilder.java#L329) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* MAJOR [The Cyclomatic Complexity of this method "resolve" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/jenkins-sonar-plugin/blob/833c3c145970cde311672d9a9bad9d879c6197cd/src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java#L45) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Pull request merged**